sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/CollationBenchmark.scala

```diff
@@ -0,0 +1,119 @@
+      "MNO", "mno", "PQR", "pqr", "STU", "stu", "VWX", "vwx", "YZ").map(UTF8String.fromString) ++
+    (18 to n).map(i => UTF8String.fromString(Random.nextString(i % 25))).sortBy(_.hashCode())
+  }
+
+  def benchmarkUTFString(collationTypes: Seq[String], utf8Strings: Seq[UTF8String]): Unit = {
+    val benchmark = collationTypes.foldLeft(
+      new Benchmark(s"collation unit benchmarks", utf8Strings.size, output = output)) {
+      (b, collationType) =>
+        b.addCase(s"equalsFunction - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.slice(0, 20).foreach(s1 =>
+            utf8Strings.filter(s =>
+            collation.equalsFunction(s, s1).booleanValue()
+            )
+          )
+        }
+        b.addCase(s"collator.compare - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.slice(0, 20).foreach(s1 =>
+            utf8Strings.sortBy(s =>
+              collation.comparator.compare(s, s1)
+            )
+          )
+        }
+        b.addCase(s"hashFunction - $collationType") { _ =>
+          val collation = CollationFactory.fetchCollation(collationType)
+          utf8Strings.map(s => collation.hashFunction.applyAsLong(s))
+        }
+        b
+    }
+    benchmark.run()
+  }
+
+  def collationBenchmarkFilterEqual(
+      collationTypes: Seq[String],
+      utf8Strings: Seq[UTF8String]): Unit = {
+    val N = 4 << 20
+    val benchmark = collationTypes.foldLeft(
+      new Benchmark(s"filter df column with collation", utf8Strings.size, output = output)) {
+      (b, collationType) =>
+        b.addCase(s"filter df column with collation - $collationType") { _ =>
+          val map: Map[String, Column] = utf8Strings.map(_.toString).zipWithIndex.map{
+            case (s, i) =>
+              (s"s${i.toString}", expr(s"collate('${s}', '$collationType')"))
+          }.toMap
+
+          val df = spark
+            .range(N)
+            .withColumn("id_s", expr("cast(id as string)"))
+            .selectExpr((Seq("id_s", "id") ++ collationTypes.map(t =>
+              s"collate(id_s, '$collationType') as k_$t")): _*)
+//            .withColumn("k_lower", expr("lower(id_s)"))
+//            .withColumn("k_upper", expr("upper(id_s)"))
+            .withColumn("s0",
+  try_element_at(array(utf8Strings.map(_.toString).map(lit): _*),
+  functions.try_add(lit(1), pmod(col("id"), lit(utf8Strings.size)).cast("int"))))
+            .withColumn("s0", expr(s"collate(s0, '$collationType')"))
+          df.where(col(s"k_$collationType") === col(s"s0"))
+                .queryExecution.executedPlan.executeCollect()
+            //          .write.mode("overwrite").format("noop").save()
+        }
+        b
+    }
+    benchmark.run()
+  }
+
+  // How to benchmark "without the rest of the spark stack"?
+
+  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
```

Comment:
Btw, thank you for doing this.
I expect that we will need to to quite some work on collation perf polishing. Having data driven benchmarks is a great way to get us to improve the perf!

Reply:
My pleasure!